Session support implementation

mshoaibijaz opened the pull request.

Commit:
Session support implementation 1st version

ummarbhutta commented on IotWeb NET45/Helper/SessionFileStorageHandler.cs:
```diff
@@ -0,0 +1,172 @@
+        }
+
+        public Task<Dictionary<SessionAttributes, object>> GetSessionMetadata(string fileName)
+        {
```
Please implement it, Even if you are not using it, put a valid implementation here

mshoaibijaz replied:
Implemented

ummarbhutta commented on IotWeb NET45/IotWeb NET45.csproj:
```diff
@@ -32,13 +32,18 @@
     <Reference Include="System.Xml" />
+    <Reference Include="Windows.Foundation.UniversalApiContract">
+      <HintPath>C:\Program Files (x86)\Windows Kits\10\References\Windows.Foundation.UniversalApiContract\2.0.0.0\Windows.Foundation.UniversalApiContract.winmd</HintPath>
+    </Reference>
```
What is this for?

mshoaibijaz replied:
Removed.

ummarbhutta commented on IotWeb Portable/Http/HttpRequestProcessor.cs:
```diff
@@ -139,9 +143,75 @@ public void ProcessHttpRequest(Stream input, Stream output)
+                    else
+                    {
+
+                    }
```
Remove the above if/else block if it is not required.

mshoaibijaz replied:
Code refactored.

ummarbhutta commented on IotWeb Portable/Http/HttpRequestProcessor.cs:
```diff
@@ -139,9 +143,75 @@ public void ProcessHttpRequest(Stream input, Stream output)
+                    else
+                    {
+
+                    }
```
Remove it as well if not required, or put a valid logic inside if/else block.

mshoaibijaz replied:
Code refactored.

ummarbhutta commented on IotWeb Portable/Http/SessionDictionary.cs:
```diff
@@ -0,0 +1,127 @@
+
+namespace IotWeb.Common.Http
+{
+    public class SessionDictionary : IDictionary<string, string>
```
Remove this class if it is not required, you can use use simple dictionary

mshoaibijaz replied:
Class has removed.

ummarbhutta commented on IotWeb UWP/Helper/SessionFileStorageHandler.cs:
```diff
@@ -0,0 +1,171 @@
+
+        public Task<Dictionary<SessionAttributes, object>> GetSessionMetadata(string fileName)
+        {
+            throw new NotImplementedException();
```
Implement this method

mshoaibijaz replied:
Implemented.

ummarbhutta reviewed (changes requested):
Please make these fixes, against my first feedback

Commit:
Session cache implementation

Commit:
SessionStorageConfiguration class refectoring

ummarbhutta commented on IotWeb Portable/Util/SessionStorageConfiguration.cs:
```diff
@@ -39,20 +33,17 @@ public SessionStorageConfiguration()
         /// <param name="sessionExpiredTime">The session expired time.</param>
-        /// <param name="sessionFileExtension">The session file extension.</param>
-        public SessionStorageConfiguration(string storagePath, int sessionExpiredTime, string sessionFileExtension)
+        public SessionStorageConfiguration(string storagePath, int sessionExpiredTime)
```
Can't we just name this class as SessionConfiguration rather than SessionStorageConfiguration?, Because `Timeout` doesn't have any relation with StorageConfiguration, but file StoragePath. Please change it a bit

ummarbhutta commented on IotWeb Portable/Util/SessionStorageConfiguration.cs:
```diff
@@ -21,13 +21,7 @@ public class SessionStorageConfiguration
         /// </summary>
         /// <value>The session expired time.</value>
         public int SessionExpiredTime { get; set; }
```
Rename it to `Timeout`

ummarbhutta commented on IotWeb Portable/Util/SessionHandler.cs:
```diff
@@ -0,0 +1,105 @@
+        }
+
+        public async Task<bool> SaveSessionData()
+        {
```
Please change access specifier of this method, I don't this it is required to be exposed

ummarbhutta commented on IotWeb Portable/Util/SessionHandler.cs:
```diff
@@ -0,0 +1,105 @@
+        }
+
+        public async Task<bool> GetSessionData(string sessionId)
+        {
```
Please change access specifier of this method, I don't this it is required to be exposed

ummarbhutta commented on IotWeb Portable/Util/SessionHandler.cs:
```diff
@@ -0,0 +1,105 @@
+        }
+
+        public async Task<bool> ClearExpiredSessions()
+        {
```
- Please change access specifier of this method, I don't this it is required to be exposed
- Secondly if method is async use Async as postfix

ummarbhutta commented on IotWeb Portable/Util/SessionHandler.cs:
```diff
@@ -0,0 +1,105 @@
+        }
+
+        public bool ClearAndDeleteSession()
+        {
```
Rename it to DestroySession

Commit:
Session cookie handling in http response

Commit:
SessionHandler class refactoring + Session expire time handling logic refactoring + HttpRequestProcessor class refactoing

Commit:
Response code "302 MovedTemporarily" addition.

Commit:
Fallback functionality in HybridSessionStorageHandler for session's data fetching + Single session deletion logic update

Commit:
Creation of New Session on logout

Commit:
SessionStorageConfiguration class renaming + Cookies update on logour

Commit:
Session handling code update, (HttpRequestProcessor, SessionHandler and HybridSessionStorageHandler class code update and fixes)

Commit:
SessionCache type changed from List to Dictionary to better handle the duplicate sessions.

UmmarIqbal committed:
some updates in comments

Commit:
HybridSessionStorageHandler fixes

Commit:
Merge branch 'session_support' of https://github.com/ummarbhutta/iotweb into session_support

Commit:
HttpServer object intialization with default session configuration + Session files storage path creation issue fixes

ummarbhutta reviewed (changes requested):
One possible issue
- There is a concept of Filters in this server, and we can use request, response and HttpContext in it, and manipulate them as well. I assume from looking into code that if someone has used after filters and have made some possible changes in session then it will not be written to session file. Please have a look in it.
- Secondly can we make session access in HttpHandler more easier than this? like ASP.NET controller? where session is available through out the controller class, and we don't have to pass it to all child function (consider of a complicated HttpRequestHandler, where we need to call many child functions to process a request and session is required in all of them)

UmmarIqbal committed:
minor changes

UmmarIqbal committed:
Merge branch 'session_support' of https://github.com/ummarbhutta/iotweb.git

Commit:
Code refactoring

Commit:
Merge branch 'session_support' of https://github.com/ummarbhutta/iotweb.git

Commit:
1- Update of Session data after handling of 'After Filters' for request.
2- HttpHandlerBase abstract class for availability of Request, Response, and Session Handler through out the inherited handler rather than passing as parameter to 'Handle Request' method.

Commit:
SessionHandler's SessionData and HybridSessionStorageHandler's SessionDataCache type update to dictionary of objects to handle complex data.

Commit:
Minor fix
